sled-agent/src/sim/http_entrypoints_storage.rs

```diff
@@ -0,0 +1,169 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! HTTP entrypoint functions for simulating the storage agent API.
+
+use dropshot::{
+    endpoint, ApiDescription, HttpError, HttpResponseDeleted, HttpResponseOk,
+    Path as TypedPath, RequestContext, TypedBody,
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+use super::storage::CrucibleData;
+
+type CrucibleAgentApiDescription = ApiDescription<Arc<CrucibleData>>;
+
+/// Returns a description of the Crucible Agent API.
+pub fn api() -> CrucibleAgentApiDescription {
+    fn register_endpoints(
+        api: &mut CrucibleAgentApiDescription,
+    ) -> Result<(), String> {
+        api.register(region_list)?;
+        api.register(region_create)?;
+        api.register(region_get)?;
+        api.register(region_delete)?;
+        Ok(())
+    }
+
+    let mut api = CrucibleAgentApiDescription::new();
+    if let Err(err) = register_endpoints(&mut api) {
+        panic!("failed to register entrypoints: {}", err);
+    }
+    api
+}
+
+// XXX XXX XXX THIS SUCKS XXX XXX XXX
+//
+// I need to re-define all structs used in the crucible agent
+// API to ensure they have the traits I need. The ones re-exported
+// through the client bindings, i.e., crucible_agent_client::types,
+// don't implement the "JsonSchema" trait, and cannot be used as
+// parameters to these Dropshot endpoints.
+//
+// I'd like them to! If we could ensure the generated client
+// also implemented e.g. JsonSchema, this might work?
+
+// To quickly test the type compatibility, uncomment the lines below,
+// and remove the hand-rolled implementations.
+
+// pub type RegionId = crucible_agent_client::types::RegionId;
+// pub type State = crucible_agent_client::types::State;
+// pub type CreateRegion = crucible_agent_client::types::CreateRegion;
+// pub type Region = crucible_agent_client::types::Region;
+// pub type RegionPath = crucible_agent_client::types::RegionPath;
+
+#[derive(
+    Serialize,
+    Deserialize,
+    JsonSchema,
+    Debug,
+    PartialEq,
+    Eq,
+    Clone,
+    PartialOrd,
+    Ord,
+)]
+pub struct RegionId(pub String);
+
+#[derive(Serialize, Deserialize, JsonSchema, Debug, PartialEq, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum State {
+    Requested,
+    Created,
+    Tombstoned,
```

Comment:
What does Tombstoned mean here?   https://www.youtube.com/watch?v=MICPyrnGYwg

Reply:
I just do what [the client bindings tell me to do](https://github.com/oxidecomputer/crucible/blob/7e8d0c314adb43698e8a65d1c55abf2225d35c2c/openapi/crucible-agent.json#L188)!

But I'd suspect it's the common interpretation [here](https://en.wikipedia.org/wiki/Tombstone_(data_store)), where it's a stepping stone on the way to deletion.